shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,313 @@
+     * be executed to ensure that the Provider can run.
+     *
+     * @tparam T the condition expression that must be true in order to execute this task.
+     */
+    template <typename State, template <typename> class expr, enum State::Value value>
+    struct When {
+
+        /**
+         * Bind a when expression in the Director.
+         *
+         * @tparam DSL the DSL from NUClear
+         *
+         * @param reaction the reaction that is having this when condition bound to it
+         */
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+
+            // Tell the director about this when condition
+            reaction->reactor.powerplant.emit(std::make_unique<commands::WhenExpression>(
+                reaction,
+                // typeindex of the enum value
+                typeid(State),
+                // Function that uses expr to determine if the passed value v is valid
+                [](const int& v) { return expr<int>()(v, value); },
+                // Function that uses get to get the current state of the reaction
+                []() { return NUClear::dsl::operation::CacheGet<State>::get(); },
+                // Binder function that lets a reactor bind a function that is called when the state changes
+                [](NUClear::Reactor& reactor,
+                   std::function<void(const int&)> fn) -> NUClear::threading::ReactionHandle {
+                    return reactor.on<NUClear::dsl::word::Trigger<State>>().then(
+                        [fn](const State& s) { fn(static_cast<int>(s)); });
+                }));
+        }
+    };
+
+    /**
+     * This DSL word is used to create a promise that at some point when running this reaction the state indicated
+     * by T will be true. It may not be on any particular run of the Provider, in which case it will continue to run
+     * that reaction when new tasks arrive until it is true, or a change elsewhere changes which Provider to run.
+     *
+     * @tparam T the condition expression that this reaction will make true before leaving.
+     */
+    template <typename State, enum State::Value value>
+    struct Causing {
+
+        /**
+         * Bind a causing expression in the Director.
+         *
+         * @tparam DSL the DSL from NUClear
+         *
+         * @param reaction the reaction that is having this when condition bound to it
+         */
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+            // Tell the director
+            reaction->reactor.powerplant.emit(
+                std::make_unique<commands::CausingExpression>(reaction, typeid(State), value));
+        }
+    };
+
+    /**
+     * A Task to be executed by the director, and a priority with which to execute this task.
```

Comment:
What indicates best priority? Higher is best?

Reply:
Higher is best, although the algorithm hasn't been written yet :P